File: frontend/src/services/avlTreeService.js
import apiClient from "./apiClient";

// Insert a new seat into the AVL tree
export const insertSeat = async (seatId, isAvailable, seatType) => {
  try {
    const response = await apiClient.post("/avltree/insert", {
      seat_id: seatId,
      is_available: isAvailable,
      seat_type: seatType,
    });
    // Return the entire data (just like seatHashMap)
    // If the backend returns {success: false, error: "..."}, we handle it on the component side
    return response.data;
  } catch (error) {
    // If a network or unexpected error occurs, handle it here
    const errMsg =
      error.response?.data?.error ||
      error.response?.data?.detail ||
      "Failed to insert seat.";
    throw new Error(errMsg);
  }
};

// Book a seat in the AVL tree
export const bookSeat = async (seatId) => {
  try {
    const response = await apiClient.post("/avltree/book", null, {
      params: { seat_id: seatId },
    });
    return response.data;
  } catch (error) {
    const errMsg =
      error.response?.data?.error ||
      error.response?.data?.detail ||
      "Failed to book seat.";
    throw new Error(errMsg);
  }
};

// Cancel a seat booking
export const cancelSeat = async (seatId) => {
  try {
    const response = await apiClient.post("/avltree/cancel", null, {
      params: { seat_id: seatId },
    });
    return response.data;
  } catch (error) {
    const errMsg =
      error.response?.data?.error ||
      error.response?.data?.detail ||
      "Failed to cancel seat.";
    throw new Error(errMsg);
  }
};

// Find the nearest available seat
export const findNearestAvailableSeat = async (seatId) => {
  try {
    const response = await apiClient.get("/avltree/nearest", {
      params: { seat_id: seatId },
    });
    return response.data;
  } catch (error) {
    const errMsg =
      error.response?.data?.error ||
      error.response?.data?.detail ||
      "Failed to find the nearest seat.";
    throw new Error(errMsg);
  }
};

File: backend/app/services/avltree_service.py
from cpp_backend import AVLTree, Seat
import asyncio

# Singleton instance of AVLTree
avl_tree = AVLTree()
lock = asyncio.Lock()

async def insert_seat(seat_id: int, is_available: bool, seat_type: str):
    async with lock:
        # Check if seat already exists before inserting
        # Use the AVLTree search method
        if avl_tree.search(seat_id):
            return {"success": False, "error": "Seat already exists"}
        seat = Seat(seat_id, is_available, seat_type)
        avl_tree.insert(seat)
    return {"success": True}

async def remove_seat(seat_id: int, is_available: bool, seat_type: str):
    async with lock:
        seat = Seat(seat_id, is_available, seat_type)
        avl_tree.remove(seat)
    return {"success": True}

async def book_seat(seat_id: int):
    async with lock:
        success = avl_tree.bookSeat(seat_id)
    return {"success": success, "error": None} if success else {"success": False, "error": "Seat not available"}

async def cancel_seat(seat_id: int):
    async with lock:
        success = avl_tree.cancelSeat(seat_id)
    return {"success": success, "error": None} if success else {"success": False, "error": "Seat not booked"}

async def find_nearest_seat(seat_id: int):
    async with lock:
        nearest_seat = avl_tree.findNearestAvailableSeat(seat_id)
    if nearest_seat:
        return {
            "seatID": nearest_seat.seatID,
            "isAvailable": nearest_seat.isAvailable,
            "seatType": nearest_seat.seatType,
        }
    return {"success": False, "error": "No available seats found"}
